File: LIS_all3.java
import java.util.*;

public class Code {
	
	public static int lowerBound(ArrayList<Integer> ar, int low, int high, int key) {
		while(low < high) {
			int mid = (low + high)/2;
			if(key == ar.get(mid))
				return mid;
			if(key > ar.get(mid)) {
				low = mid+1;
			}else {
				high = mid;
			}
		}
		return low;
	}
	
	public static int LISBF(int [] ar) {
		int n=ar.length;
		if(n <= 0)
			return 0;
		int res=1;
		for(int i=0; i<(1<<n); i++) {
			boolean firstTime=true;
			int curr=1;
			String bits=Integer.toBinaryString(i);
			int prev=-1;
			for(int j=0; j<bits.length() && j<ar.length; j++) {
				if(bits.charAt(j) == '1') {
					if(firstTime) {
						firstTime=false;
						prev=ar[j];
					}else {
						if(ar[j] <= prev)
							break;
						else {
							prev=ar[j];
							curr++;
						}
					}
				}
			}
			res = Math.max(curr, res);
		}
		return res;
	}
	
	public static int LISDP(int [] ar) {
		int n=ar.length;
		if(n <= 0)
			return 0;
		int res=1;
		int [] lisLengthUpto = new int [n];
		Arrays.fill(lisLengthUpto, 1);
		for(int i=1; i<n; i++) {
			int bestprefixLength=0;
			for(int j=0; j<i; j++) {
				if(ar[j] < ar[i]) {
					bestprefixLength = Math.max(bestprefixLength, lisLengthUpto[j]);
				}
			}
			lisLengthUpto[i] = Math.max(bestprefixLength+1, lisLengthUpto[i]);
			res = Math.max(res, lisLengthUpto[i]);
		}
		return res;
	}
	
	public static int LISUBLB(int [] ar) {
		int n=ar.length;
		if(n <= 0)
			return 0;
		int res=1;
		ArrayList<Integer> sortedlist = new ArrayList<Integer>();
		for(int i=0; i<n; i++) {
			if(sortedlist.isEmpty()) {
				sortedlist.add(ar[i]);
				continue;
			}
			if(ar[i] > sortedlist.get(sortedlist.size() - 1)) {
				sortedlist.add(ar[i]);
				res++;
			}else {
				if(sortedlist.get(lowerBound(sortedlist, 0, sortedlist.size() - 1, ar[i])) == ar[i])
					continue;
				sortedlist.set(lowerBound(sortedlist, 0, sortedlist.size() - 1, ar[i]), ar[i]);
			}
		}
		return res;
	}
	
	public static void main(String[] args) {
		Scanner sc=new Scanner(System.in);
		System.out.println("Enter length of array: ");
		int n=sc.nextInt();
		System.out.println("Enter elements of array: ");
		int [] ar=new int[n];
		for(int i=0; i<n; i++)
			ar[i] = sc.nextInt();
		System.out.println("LIS using Brute force: "+LISBF(ar));
		System.out.println("LIS using DP: "+LISDP(ar));
		System.out.println("LIS using UBLB: "+LISUBLB(ar));
	}
}

File: LIS_all3.cpp
#include<bits/stdc++.h>
using namespace std;

#define ll long long int

int LISBF(int * ar, int n) {
	if(n <= 0)
	       return 0;
	int res=1;
	for(int i=0; i<(1<<n); i++) {
		int curr=1;
		int prev=-1;
		bool firstTime=true;
		string bits=bitset<10>(i).to_string();
		for(int j=0; j<bits.size() and j<n; j++) {
			if(bits[j] == '1')
			{
				if(firstTime) {
					prev=ar[j];
					firstTime=false;
				}
				else {
					if(prev < ar[j]) {
						curr++;
						prev=ar[j];
					}else {
						break;
					}
				}
			}
		}
		res = max(res, curr);
	}
	return res;
}

int LISDP(int * ar, int n) {
	if(n<=0)
		return 0;
	int res=1;
	int * LISupto = new int[n];
	for(int i=0; i<n; i++)
		LISupto[i] = 1;
	for(int i=1; i<n; i++) {
		int bestPrefixLength=0;
		for(int j=0; j<i; j++) {
			if(ar[j] < ar[i]) {
				bestPrefixLength = max(bestPrefixLength, LISupto[j]);
			}
		}
		LISupto[i] = max(LISupto[i], bestPrefixLength + 1);
		res = max(res, LISupto[i]);
	}
	return res;
}

int LISUBLB(int * ar, int n) {
	if(n <= 0 )
		return 0;
	int res=1;
	vector<int> sortedList;
	for(int i=0; i<n; i++) {
		if(sortedList.empty()){
			sortedList.push_back(ar[i]);
			continue;
		}
		if(ar[i] > sortedList.back()) {
			sortedList.push_back(ar[i]);
			res++;
		}
		else if(sortedList[lower_bound(sortedList.begin(), sortedList.end(), ar[i]) - sortedList.begin()] == ar[i]){
			continue;
		}
		else {
			sortedList[lower_bound(sortedList.begin(), sortedList.end(), ar[i]) - sortedList.begin()] = ar[i];
		}
	}
	return res;
}

int main(){
	cout << "Enter length: ";
	int n;
	cin >> n;
	cout << "Enter elements: ";
	int * ar=new int[n];
	for(int i=0; i<n; i++)
		cin >> ar[i];
	cout << "\nLIS using Brute force: " << LISBF(ar, n);
	cout << "\nLIS using DP: " << LISDP(ar, n);
	cout << "\nLIS using UBLB: " << LISUBLB(ar, n);
	cout << '\n';
	return 0;
}